File: examples/csharp/Inventory/src/InventoryService.cs
using Angzarr;
using Examples;
using Grpc.Core;
using Serilog;

namespace Angzarr.Examples.Inventory;

public class InventoryService : BusinessLogic.BusinessLogicBase
{
    private readonly IInventoryLogic _logic;
    private readonly Serilog.ILogger _logger;

    public InventoryService(IInventoryLogic logic, Serilog.ILogger logger)
    {
        _logic = logic;
        _logger = logger.ForContext<InventoryService>();
    }

    public override Task<BusinessResponse> Handle(ContextualCommand request, ServerCallContext context)
    {
        try
        {
            var events = ProcessCommand(request);
            var response = new BusinessResponse { Events = events };
            return Task.FromResult(response);
        }
        catch (CommandValidationException ex)
        {
            throw new RpcException(new Status(ex.StatusCode, ex.Message));
        }
        catch (Exception ex)
        {
            _logger.Error(ex, "Unexpected error processing command");
            throw new RpcException(new Status(StatusCode.Internal, $"Internal error: {ex.Message}"));
        }
    }

    private EventBook ProcessCommand(ContextualCommand request)
    {
        var cmdBook = request.Command;
        var priorEvents = request.Events;

        if (cmdBook == null || cmdBook.Pages.Count == 0)
            throw CommandValidationException.InvalidArgument("CommandBook has no pages");

        var cmdPage = cmdBook.Pages[0];
        if (cmdPage.Command == null)
            throw CommandValidationException.InvalidArgument("Command page has no command");

        var state = _logic.RebuildState(priorEvents);
        var command = cmdPage.Command;
        var typeUrl = command.TypeUrl;

        EventBook result;

        if (typeUrl.EndsWith("InitializeStock"))
        {
            var cmd = command.Unpack<InitializeStock>();
            result = _logic.HandleInitializeStock(state, cmd.ProductId, cmd.Quantity);
        }
        else if (typeUrl.EndsWith("ReceiveStock"))
        {
            var cmd = command.Unpack<ReceiveStock>();
            result = _logic.HandleReceiveStock(state, cmd.Quantity);
        }
        else if (typeUrl.EndsWith("ReserveStock"))
        {
            var cmd = command.Unpack<ReserveStock>();
            result = _logic.HandleReserveStock(state, cmd.OrderId, cmd.Quantity);
        }
        else if (typeUrl.EndsWith("ReleaseReservation"))
        {
            var cmd = command.Unpack<ReleaseReservation>();
            result = _logic.HandleReleaseReservation(state, cmd.OrderId);
        }
        else if (typeUrl.EndsWith("CommitReservation"))
        {
            var cmd = command.Unpack<CommitReservation>();
            result = _logic.HandleCommitReservation(state, cmd.OrderId);
        }
        else
        {
            throw CommandValidationException.InvalidArgument($"Unknown command type: {typeUrl}");
        }

        result.Cover = cmdBook.Cover;
        return result;
    }
}

File: examples/cpp/inventory/src/inventory_service.cc
#include "inventory_logic.hpp"
#include "logging.hpp"
#include "validation_error.hpp"
#include "angzarr.grpc.pb.h"
#include <grpcpp/grpcpp.h>

namespace inventory {

class InventoryService final : public angzarr::BusinessLogic::Service {
public:
    grpc::Status Handle(grpc::ServerContext* context,
                        const angzarr::ContextualCommand* request,
                        angzarr::BusinessResponse* response) override {
        try {
            const auto& cmd_book = request->command();
            const auto* prior_events = request->has_events() ? &request->events() : nullptr;

            if (cmd_book.pages().empty()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "CommandBook has no pages");
            }

            const auto& cmd_page = cmd_book.pages(0);
            if (!cmd_page.has_command()) {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Command page has no command");
            }

            auto state = InventoryLogic::rebuild_state(prior_events);
            const auto& type_url = cmd_page.command().type_url();

            google::protobuf::Any event_any;

            if (type_url.find("InitializeStock") != std::string::npos) {
                examples::InitializeStock cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("inventory", "initializing_stock",
                    {{"product_id", cmd.product_id()}, {"quantity", cmd.initial_quantity()}});
                auto event = InventoryLogic::handle_initialize_stock(
                    state, cmd.product_id(), cmd.initial_quantity());
                event_any.PackFrom(event);
            } else if (type_url.find("ReceiveStock") != std::string::npos) {
                examples::ReceiveStock cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("inventory", "receiving_stock", {{"quantity", cmd.quantity()}});
                auto event = InventoryLogic::handle_receive_stock(
                    state, cmd.quantity(), cmd.reference());
                event_any.PackFrom(event);
            } else if (type_url.find("ReserveStock") != std::string::npos) {
                examples::ReserveStock cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("inventory", "reserving_stock",
                    {{"order_id", cmd.order_id()}, {"quantity", cmd.quantity()}});
                auto event = InventoryLogic::handle_reserve_stock(
                    state, cmd.order_id(), cmd.quantity());
                event_any.PackFrom(event);
            } else if (type_url.find("ReleaseReservation") != std::string::npos) {
                examples::ReleaseReservation cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("inventory", "releasing_reservation", {{"order_id", cmd.order_id()}});
                auto event = InventoryLogic::handle_release_reservation(state, cmd.order_id());
                event_any.PackFrom(event);
            } else if (type_url.find("CommitReservation") != std::string::npos) {
                examples::CommitReservation cmd;
                cmd_page.command().UnpackTo(&cmd);
                angzarr::log_info("inventory", "committing_reservation", {{"order_id", cmd.order_id()}});
                auto event = InventoryLogic::handle_commit_reservation(state, cmd.order_id());
                event_any.PackFrom(event);
            } else {
                return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Unknown command type: " + type_url);
            }

            auto* event_book = response->mutable_events();
            if (cmd_book.has_cover()) *event_book->mutable_cover() = cmd_book.cover();

            auto* event_page = event_book->add_pages();
            event_page->set_num(0);
            *event_page->mutable_event() = event_any;
            event_page->mutable_created_at()->set_seconds(
                std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));

            return grpc::Status::OK;
        } catch (const angzarr::ValidationError& e) {
            return e.to_grpc_status();
        }
    }
};

std::unique_ptr<angzarr::BusinessLogic::Service> create_inventory_service() {
    return std::make_unique<InventoryService>();
}

}  // namespace inventory
